Formula/b/beakerlib.rb

```diff
@@ -0,0 +1,39 @@
+class Beakerlib < Formula
+  desc "Shell-level integration testing library"
+  homepage "https://github.com/beakerlib/beakerlib"
+  url "https://github.com/beakerlib/beakerlib/archive/refs/tags/1.30.tar.gz"
+  sha256 "bd06fc61b32d9caf4324587706a8363e37e771355da8297d0c5ba0023ae31098"
+  license "GPL-2.0-only"
+
+  on_macos do
+    # Fix `readlink`
+    depends_on "coreutils"
+    depends_on "gnu-getopt"
+  end
+
+  def install
+    system "make", "DD=#{prefix}", "install"
+    (prefix.glob "**/*.sh").each do |f|
+      inreplace f, "readlink", "#{Formula["coreutils"].opt_bin}/greadlink", false if OS.mac?
+      inreplace f, "getopt", "#{Formula["gnu-getopt"].opt_bin}/getopt", false if OS.mac?
```

Comment:
This feels like it should be an option in the upstream `Makefile` rather than requiring us to patch it. Even better would be to avoid needing these GNU dependencies on macOS.

Reply:
Unfortunately that is tricky because this package is a `bash` based testing-framework so it uses a lot of linux specific commands, not only for the build, but for the runtime execution as well, i.e. the installed `.sh` files make use of extended `getopt` functions. This is also complicated by the fact that this must run consistently after reboot as well.

The replacement is overzealous, in that it replaces comments as well, but as for the functionality, it should work more reliably with such replacement.

Reply:
As-is: this software does not seem to be ported to macOS.

Reply:
> As-is: this software does not seem to be ported to macOS.

Could you elaborate? Of course the functionality is not completely equivalent, but as long as the basic functionality like the `rlAssert` and journaling work, that is already a good enough first step for the `tmt` package integration. That I am already testing in the smoke-test included.

Having a reference formula would also help upstream expand the support of MacOS.

Reply:
> Could you elaborate?

If the software requires `inreplace` like this to work on macOS: it is not ported to macOS.

You could argue further that if it requires GNU tools like this it's not ported but: that's not blocking this PR.

Reply:
Looking at the code implementation: https://github.com/search?q=repo%3Abeakerlib%2Fbeakerlib%20getopt&type=code

would you have a suggestion on how to make it macOS portable?

Reply:
> I am not sure what you mean here, the upstream project does various detections to check for GNU vs BSD `getopt`

Why does the formula need to depend on/use `gnu-getopt` then?

Reply:
Because the runtime shell scripts that are used by the user, e.g. `source #{share}/beakerlib/beakerlib.sh`, they **require** enhanced `getopt`. The checks inside the script check that `getopt` is enhanced, and if that fails, the whole script throws an error message. This happens both at the install stage, where some of these scripts are used internally, as well as at runtime when these are used by the user. So to get the **enhanced** `getopt` the `gnu-getopt` formula is needed with replacement of the `getopt` taken from `PATH` with the fixed absolute path to `gnu-getopt`

Reply:
Ok, I see, thanks.

My numbered list from above still applies here.

Reply:
I can imagine it to be a part of the installation process in the Makefile (some `sed` processing).
I would rather keep the shell scripts untouched as there would be higher chance of direct usage in future development which would lead to a need of subsequent fixes afterwards.

Reply:
> @Bo98
> A PATH wrapper/configuration seems like the simplest approach here?

If you mean adding to `PATH` for the `.rb` formula, that wouldn't work because it needs to be made available after the installation for users that call beakerlib via `source .../beakerlib.sh`.

If you mean altering `PATH` within the scripts, we are just as well back to square one with the `sed`/`inreplace`

> @sopos 
> I can imagine it to be a part of the installation process in the Makefile (some sed processing).
> I would rather keep the shell scripts untouched as there would be higher chance of direct usage in future development which would lead to a need of subsequent fixes afterwards.

The idea for this is to add a configurable redirection to the actual `getopt` command, that will be `sed`-ed by homebrew, and otherwise left intact in order for the scripts to be executable in place, e.g. for the `logging.sh`
```patch
echo "${__INTERNAL_SOURCED}" | grep -qF -- " ${BASH_SOURCE} " && return || __INTERNAL_SOURCED+=" ${BASH_SOURCE} "

+ declare GETOPT_CMD=getopt

export __INTERNAL_DEFAULT_SUBMIT_LOG=__INTERNAL_FileSubmit

...

rlFileSubmit() {
-    GETOPT=$(getopt -o s: -- "$@" 2> >(while read -r line; do rlLogError "$FUNCNAME: $line"; done))
+    GETOPT=$(GETOPT_CMD -o s: -- "$@" 2> >(while read -r line; do rlLogError "$FUNCNAME: $line"; done))
    eval set -- "$GETOPT"
```
It would require yet another release to be accepted for homebrew, but we can experiment on the tap with patches to see if everything works fine

@SMillerDev @MikeMcQuaid We would appreciate accumulating some more feedback so that we don't have to cut new releases upstream this often.

As far as the `1.30` release is concerned, the `inreplace` does not do anything surprising (assuming it is correctly case-sensitive, which can be quickly sanity checked from the bottles), and it can be later refined on later release

Reply:
> If you mean altering PATH within the scripts, we are just as well back to square one with the sed/inreplace

Is there multiple entry points for the scripts or does everything start with `beakerlib.sh`? If it's the latter then wrapping that with a PATH configuration is simple enough.

Reply:
> I can imagine it to be a part of the installation process in the Makefile (some sed processing).

This makes sense to me too 👍🏻 

> It would require yet another release to be accepted for homebrew, but we can experiment on the tap with patches to see if everything works fine

This can be merged before the patches are in a release as long as they are merged upstream first.

Reply:
> This can be merged before the patches are in a release as long as they are merged upstream first.

The beakerlib upstream release quite cheap operation. If you provide reasonable PR, it can be merged and v1.30.1 can be done in a heart beat.

Reply:
Adding the `sed` replacement within the upstream `Makefile`:
- PR : https://github.com/beakerlib/beakerlib/pull/172
- Run in third-party tap: https://github.com/LecrisUT/homebrew-tmt/pull/1

Reply:
@MikeMcQuaid The upstream patches were merged how do you want to proceed with this?

Reply:
Thanks @LecrisUT! Apply the merged PR with a `patch do` block here and comment saying what it does and when it can be removed and adapt the `install` accordingly.

Reply:
:+1: How do I deal with the CI breaking down and not continuing to the test stage after it detects a `patch`? Can I link `patch do` to the [PR patch url](https://patch-diff.githubusercontent.com/raw/beakerlib/beakerlib/pull/172.patch)?

Reply:
Still can't get around it ><:
```
==> FAILED
Full audit --formula beakerlib --online --new output
  beakerlib
    * Formulae should not require patches to build. Patches should be submitted and accepted upstream first.
  Error: 1 problem in 1 formula detected.
```

Reply:
Yes. Check how other formulae use patches/PR URLs.

Reply:
@LecrisUT That's fine, we can ignore/override that.

Reply:
Well that would need to be done in this PR somehow in order to run the CI for all macOS variants to confirm that everything is working